staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -264,6 +277,8 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		s.GenericAPIServer.AuditBackend.Shutdown()
```

Comment:
can the audit shutdown be a pre stop hook as well?

Reply:
/cc @hzxuzhonghu @crassirostris

Reply:
Yes, audit backend can register a prestop/poststop hook too.

Reply:
I think we should even have a synchronization of closing the listener and calling the hooks here.

Reply:
This of course means that some API server clients in an HA setup might get a connection refused. On the other hand, we cannot make kube-proxy updates synchronous anyway. So it does not matter much for external clients which order we have here.

But, auditing needs this synchronization.

Reply:
You mean do not call the shutdown hooks until the listener is closed?

Reply:
Yes, but I fine with wiring this in a follow-up. We definitely need that for auditing.